### DISCUSSION TITLE
Quick (nube) question on Django vector searching...

### DISCUSSION THREAD

**Discussion by phlashworth** [pkavumba/django-vectordb#3] [Q&A] [answered]
I'm looking to bring in vector search to one of my Django projects and I'm not clear on the best (most appropriate) option to do this... I stumbled upon this repo and it looks great! My project uses Postresql and I know there is already Searchvector, searchquery and searchrank support for postgres [here](https://docs.djangoproject.com/en/4.2/ref/contrib/postgres/search/) 

Will that do what most of what this repo provides (except choice of model for embedding) **IF** using postgres or are there some extra benefits/features/performance gains to be had in using this library? I'm keen to bake in flexibility to add richer features to my project in future so, although I'm keen not to use a sledge hammer to crack a nut, I'm open to use this if there is a good case for doing so.

Lastly, does this store the embeddings for strings in the project's native DB or does it use something else?

Many thanks in advance!

**Comment by pkavumba** [comment, marked as answer]
Hello @phlashworth,

I'm glad to hear that you're considering using the django-vectordb for your Django project. Let's dive into your questions.

Firstly, let's distinguish between full-text search and semantic search through vector embeddings. Full-text search, as provided by PostgreSQL's Searchvector, Searchquery, and Searchrank, is a traditional method of searching. For instance, if you search for "apple", it will return all documents containing the word "apple". It's a simple and straightforward method, but it doesn't understand the context or meaning of the words. So, if you search for "apple", it won't return documents that mention "fruit" or "iPhone", even though these are related concepts.

On the other hand, semantic search through django-vectordb is a more advanced method. It involves converting words or phrases or documents into vectors (arrays of numbers) that capture their meanings. For example, the vector for "apple" might be close to the vectors for "fruit" and "iPhone", because these are related concepts. This means that a semantic search for "apple" could return documents that mention "fruit" or "iPhone", even if they don't contain the word "apple". This makes the search more powerful and flexible, as it can find relevant results even if they don't use the exact same words as the search terms.

Now, let's talk about the benefits of using the django-vectordb library. While PostgreSQL's full-text search features are powerful, they don't provide the semantic understanding that vector embeddings do. By using django-vectordb, you can add this semantic understanding to your search, making it more flexible and powerful. This can be particularly useful if your project involves complex or ambiguous search terms, or if you want to provide more relevant results to your users. Moreover, django-vectordb can be utilized to identify related posts/objects, enabling you to effortlessly incorporate recommendations, which may improve user engagement and overall user experience on your platform.

In addition, django-vectordb is a flexible tool. Even though it provides good default settings, it also lets you choose different models for creating the embeddings. This means you can pick the model that works best for your needs. It could be a general model that's been trained on lots of different text, or a specialized model that's been trained on a specific type of text. This flexibility can be really helpful if your project involves specific types of text that aren't well-handled by general models.

As for where the embeddings are stored, django-vectordb stores them in your project's native database. This means you don't need to set up and manage a separate database for the embeddings, simplifying your project's architecture and potentially reducing costs.

In conclusion, while PostgreSQL's full-text search features are powerful and may be sufficient for some projects, django-vectordb provides additional flexibility and power through the use of vector embeddings. If you're looking to add richer features and recommendations to your project, it could be a good choice.

I hope this helps! If you have any further questions, please don't hesitate to ask.

**Reply by phlashworth** [reply]
Thank you so much @pkavumba for the comprehensive answer! I think my confusion about PostgreSQL's capabilities stem from this [documentation](https://www.postgresql.org/docs/current/textsearch-controls.html#TEXTSEARCH-RANKING) - see the last search result in the example where neither of the terms 'neutrino' or 'dark matter' appear in the result. PG seems to support a tsvector data type for this purpose but I have no understanding of how flexible it is or whether it brings about any efficiency gains over alternative approaches. 

What I like most about django-vectordb (in addition to the flexibility etc discussed above) is that this will work on DBs other than just Postgres which means it'll be suitable for some of my other projects too which use other DBs. I may reach out to you on DM to discuss those in due course if that's ok with you.

**Reply by pkavumba** [reply]
I'm glad to hear that you found the previous explanation helpful! Allow me to dig into the result of the example in the [documentation](https://www.postgresql.org/docs/current/textsearch-controls.html#TEXTSEARCH-RANKING).

Firstly, let's clarify the difference in `vector` usage. In PostgreSQL, the `tsvector` value is a sorted list of distinct lexemes, which are words that have been normalized to merge different variants of the same word. Optionally, integer positions can be attached to lexemes. For instance, the `tsvector` for the sentence "a fat cat sat on a mat - it ate a fat rats" is represented as `'ate':9 'cat':3 'fat':2,11 'mat':7 'rat':12 'sat':4`. You'll notice that 'rats' is stored as 'rat'. The same normalization is applied to the query. Therefore, searching for 'rats' will match even documents that contain 'rat' due to this lexeme normalization. However, it doesn't understand the meaning of the words, so it can't match synonyms or related concepts like django-vectordb can. Now, rankings such as `ts_rank` or `ts_rank_cd` will only match documents that share common lexemes. Documents without any common lexemes will score 0. We can quickly verify this in PG by executing the following command in the shell:
```sh
psql -d postgres
```
If we execute the example provided in the [documentation](https://www.postgresql.org/docs/current/textsearch-controls.html#TEXTSEARCH-RANKING), we can observe that it returns a rank of zero as shown below. 
```sql
postgres=# SELECT ts_rank_cd(
  to_tsvector('english', 'Weak Lensing Distorts the Universe'),
  to_tsquery('english','neutrino|(dark & matter)')
);
 ts_rank_cd
------------
          0
(1 row)
```

Now, if you compare this to the results in the documentation, it shows a score of 0.818218. The discrepancy could be because the `textsearch` field used might be a combination of the title and the description, but this information isn't provided. However, the printout only displays the title. Thus, we can infer that the matching text that results in a rank greater than zero isn't part of the title.

```sql
SELECT title, ts_rank_cd(textsearch, query) AS rank
FROM apod, to_tsquery('neutrino|(dark & matter)') query
WHERE query @@ textsearch
ORDER BY rank DESC
LIMIT 10;
                     title                     |   rank
-----------------------------------------------+----------
 Neutrinos in the Sun                          |      3.1
 The Sudbury Neutrino Detector                 |      2.4
 A MACHO View of Galactic Dark Matter          |  2.01317
 Hot Gas and Dark Matter                       |  1.91171
 The Virgo Cluster: Hot Plasma and Dark Matter |  1.90953
 Rafting for Solar Neutrinos                   |      1.9
 NGC 4650A: Strange Galaxy and Dark Matter     |  1.85774
 Hot Gas and Dark Matter                       |   1.6123
 Ice Fishing for Cosmic Neutrinos              |      1.6
 Weak Lensing Distorts the Universe            | 0.818218
```

In contrast, django-vectordb employs a different approach by storing word or contextual embeddings. These are essentially vectors, or arrays of numbers, that encapsulate the meaning of words or phrases within their specific context. For instance, the embedding vector for the phrase "a fat cat sat on a mat - it ate a fat rats" might look something like this: `[ 0.08478471,  0.01531179, -0.00973194, ...,  0.0972335, 0.02503153,  0.08958683]`.  This vector representation is not just a simple list of words or lexemes, but a complex mathematical model that captures the semantic meaning and relationships between words. 
This allows django-vectordb to understand the meaning of words and their context, making the search more powerful and flexible.


I'm glad to hear that you're considering using django-vectordb for your other projects as well. Feel free to reach out to me via DM to discuss those projects whenever you're ready. I'm always here to help!

If you have any further questions, please don't hesitate to ask.

**Reply by phlashworth** [reply]
Brilliant, thank you again for taking the time to write this full explanation! I've just got back to work after annual leave so I'll discuss django-vectordb with the team and get their thoughts.
